linkschkr.go

```diff
@@ -0,0 +1,127 @@
+package linkschkr
+
+import (
+	"fmt"
+	"io"
+	"log"
+	"net/http"
+	"strings"
+
+	"golang.org/x/net/html"
+)
+
+type Work struct {
+	site   string
+	result chan *Result
+}
+
+type Result struct {
+	site      string
+	up        bool
+	extraURIs []string
+	workerID  int
+}
+
+func Crawler(ID int, c <-chan *Work) {
+	for {
+		work := <-c
+		result := &Result{
+			site:     work.site,
+			up:       true,
+			workerID: ID,
+		}
+		var resp *http.Response
+		var err error
+		// Avoid to download binaries
+		switch {
+		case strings.HasSuffix(work.site, ".gz"):
+			resp, err = http.Head(work.site)
+		case strings.HasSuffix(work.site, ".bz2"):
+			resp, err = http.Head(work.site)
+		case strings.HasSuffix(work.site, ".msi"):
+			resp, err = http.Head(work.site)
+		case strings.HasSuffix(work.site, ".zip"):
+			resp, err = http.Head(work.site)
+		case strings.HasSuffix(work.site, ".pkg"):
+			resp, err = http.Head(work.site)
+		default:
+			resp, err = http.Get(work.site)
+		}
+		if err != nil || resp.StatusCode != 200 {
+			result.up = false
+			work.result <- result
+			continue
+		}
+
+		extraURIs := ParseHREF(resp.Body)
+		result.extraURIs = extraURIs
+		work.result <- result
+	}
+}
+
+func ParseHREF(r io.Reader) []string {
+	URIs := []string{}
+
+	doc, err := html.Parse(r)
+	if err != nil {
+		log.Fatal(err)
+	}
+	var f func(*html.Node)
+	f = func(n *html.Node) {
+		if n.Type == html.ElementNode && n.Data == "a" {
+			for _, a := range n.Attr {
+				if a.Key == "href" {
+					if strings.HasPrefix(a.Val, "/") {
+						URIs = append(URIs, a.Val)
+					}
+
+				}
+			}
+		}
+		for c := n.FirstChild; c != nil; c = c.NextSibling {
+			f(c)
+		}
+	}
+	f(doc)
+
+	return URIs
+}
+
+func SendWork(s string, work chan *Work, results chan *Result) {
+	work <- &Work{
+		site:   s,
+		result: results,
+	}
+}
+
+func Run(nWorkers int, sites []string) {
+	work := make(chan *Work)
+	results := make(chan *Result)
+
+	for n := 0; n < nWorkers; n++ {
+		go Crawler(n, work)
+	}
+	alreadyChecked := map[string]struct{}{}
+	for _, s := range sites {
+		alreadyChecked[s] = struct{}{}
+		go SendWork(s, work, results)
+	}
+	for {
+		v := <-results
+		alreadyChecked[v.site] = struct{}{}
+		up := "up"
+		if !v.up {
+			up = "down"
+		}
+		for _, u := range v.extraURIs {
+			url := strings.Split(v.site, "/")
+			s := fmt.Sprintf("%s//%s%s", url[0], url[2], u)
+			_, ok := alreadyChecked[s]
+			if !ok {
```

Comment:
```suggestion
			if !alreadyChecked[s] {
```

...is possible if we use a map of `bool` instead of `struct{}`.